06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/MeetingScheduler/meetingscheduler.spec

```diff
@@ -0,0 +1,208 @@
+methods {
+    getStateById(uint256) returns (uint8) envfree
+    getStartTimeById(uint256) returns (uint256) envfree
+    getEndTimeById(uint256) returns (uint256) envfree
+    getNumOfParticipents(uint256) returns (uint256) envfree
+}
+
+definition uninitialized(uint256 meetingId) returns bool = 
+        getStartTimeById(meetingId) == 0 &&
+        getEndTimeById(meetingId) == 0 &&
+        getNumOfParticipents(meetingId) == 0 &&
+        getStateById(meetingId) == 0;
+
+definition initialized(uint256 meetingId) returns bool =
+        getStartTimeById(meetingId) != 0 &&
+        getEndTimeById(meetingId) != 0 &&
+        getStateById(meetingId) != 0;
+
+definition UNINITIALIZED() returns uint256 = 0;
+
+function callFunctionWithParams(method f, uint256 meetingId, uint256 startTime, uint256 endTime) {
+    env e;
+
+    if (f.selector == scheduleMeeting(uint256,uint256,uint256).selector) {
+        scheduleMeeting(e, meetingId, startTime, endTime);
+    } else if (f.selector == startMeeting(uint256).selector) {
+        startMeeting(e, meetingId);
+    } else if (f.selector == cancelMeeting(uint256).selector) {
+        cancelMeeting(e, meetingId);
+    } else if (f.selector == endMeeting(uint256).selector) {
+        endMeeting(e, meetingId);
+    } else if (f.selector == joinMeeting(uint256).selector) {
+        joinMeeting(e, meetingId);
+    } else {
+        calldataarg args;
+        f(e, args);
+    }
+}
+
+
+// TODO: I don't know how to enforce that the parametric function calls each need to use a specific
+// meetingId. Gotta ask Mike
+// any 2 functions called on 2 different meetingId's will have no effect on each other's meetings
+// rule aChangeToOneMeetingDoesNotAffectTheOther(uint256 meetingId1, uint256 meetingId2) {
+//     env e1; calldataarg args1;
+//     env e2; calldataarg args2;
+//     method f1;
+//     method f2;
+
+//     // call the parametric function on two different meetings
+//     require meetingId1 != meetingId2;
+
+//     // get the state of the 2 meetings before the calls
+//     uint8 status1Before; uint256 startTime1Before; uint256 endTime1Before; uint256 numParticipants1Before;
+//     status1Before = getStateById(meetingId1);
+//     startTime1Before = getStartTimeById(meetingId1);
+//     endTime1Before = getEndTimeById(meetingId1);
+//     numParticipants1Before = getNumOfParticipents(meetingId1);
+
+//     uint8 status2Before; uint256 startTime2Before; uint256 endTime2Before; uint256 numParticipants2Before;
+//     status2Before = getStateById(meetingId2);
+//     startTime2Before = getStartTimeById(meetingId2);
+//     endTime2Before = getEndTimeById(meetingId2);
+//     numParticipants2Before = getNumOfParticipents(meetingId2);
+
+
+//     f1(e1, args1);
+
+//     // TODO
+
+//     assert true;
+
+// }
+
+
+// TODO: i don't know how to compare reachability of 2 function calls (1 by the organizer, the other
```

Comment:
@tadams do you know how to do this? Read the rule description below for what I'm trying to do.

Reply:
Here is how this rule can be implemented: 

```
rule nonOrganizerCallsAlwaysSucceedForOrganizer(method f, address nonOrganizer, address organizer) {
  env e1; env2; calldataarg args;
  require e1.msg.sender == nonOrganizer;
  require e2.msg.sender == organizer;
  require e1.msg.vale == e2.msg.value;
  // ... (all other values of the env must be the same)
  storage initial = lastStorage;

  f(e1, args);
  bool nonOrgCallSuccess = !lastReverted;

  f(e2, args) at initial;
  bool orgCallSuccess = !lastReverted;

  assert nonOrgCallSuccess => orgCallSuccess, "A function call that a non-organizer made successfully must also be callable by the organizer";
}
```

Reply:
done